### ISSUE TITLE
Cached results issue on rescan [tls.imirhil.fr]

### ISSUE THREAD

**Issue by FlorentATo** [mozilla/http-observatory-website#60]
Consider https://tls.imirhil.fr/ (awesome job @aeris !)
A given domain can be tested only once every hour and the result (https://tls.imirhil.fr/https/example.com.json) is cached until the domain is checked again.

So even though a user hits "Initiate rescan", the tool doesn't call imirhil's refresh API to reset the scan (i.e. https://tls.imirhil.fr/https/example.com/refresh), thus returning potentially inconsistent results.

Suggestion: 

- Case 1: No data cached -> Return the result + current timestamp
- Case 2: Data in cache && Tcurrent - Ttest < 1h: Return cached data + cache timestamp
- Case 3: Data in cache && Tcurrent - Ttest >= 1h: Reset data and goto case 1.

tl;dr: Always try to refresh the cache :)

**Comment by marumari** [comment]
Due to concerns about load on our 3rd party providers, this isn't something we can fix.  We do talk about it in the FAQ though:

https://observatory.mozilla.org/faq.html

Under third-party integration at the bottom.  :)

**Comment by FlorentATo** [comment]
We can't fix it, I understand.

But if a domain has already been tested (either through imirhil.fr or the observatory), the returned result will **never** be refreshed, no matter how hard I hit the rescan button.
In other words, for this particular third-party, the test is valid **only** if the domain has never been previously tested.

I'm pretty confident this issue only occurs with this third-party in particular. Shouldn't we find a way to improve that ?

**Comment by marumari** [comment]
@aeris, how would you like to handle this?  We currently never force rescans on third party sites.  We either could force a rescan, or maybe you'd be able to update something such that if the previous scan was >24h old, it scans again?

**Comment by aeris** [comment]
Don’t know how you cache on your side, but I have the date of the analysis on the JSON.
If you detect a date change, use the new data :)
API consultation is pretty light on my side, I don’t expect trouble if you call it more often.

Only TLS analysis is heavy, and can be very problematic if redo too much. Tested server take a **huge** load of TLS handshake, and service owner are generally not very happy, thinking about DoS or attack if not the initiator of the scan….
This is why I have a 1h refresh delay, as far as I’m not able to find a way for the owner to notify me he really wants to be rechecked (currently thinking about a HTTP header or a DNS entry).

If you want to force an overall rescan, better to exclude TLS to avoid DoS on a server for only a CSP change check /o\

**Comment by discountufoparts** [comment]
A particular site I am testing appears to be stuck returning the cached results.  CSP is implemented and I am receiving an A+ from securityheaders.io  but Observatory continues to fail the site.

**Comment by marumari** [comment]
In the top section or for a third party?  Can you provide the site name?

**Comment by discountufoparts** [comment]
ksw.timberwolfnewmedia.com  I have tested other sites and they all appear
to function normally.

**Comment by marumari** [comment]
The Observatory looks at headers returned over https if the site is available that way.  securityheaders.io returns the same failing grade:

https://securityheaders.io/?q=https%3A%2F%2Fksw.timberwolfnewmedia.com%2F

If you ran a local scan, disabling HTTPS, it would show the same A+, but because ksw.timberwolfmedia.com responds over :443 (even if for a different site), it produces that failing grade.
